File: merge-tonic-lib/tonic/hydration/conflict_scan.py
"""Conflict marker scan (parity with conflictScan.ts semantics)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from tonic.git_conflict_parser import parse_git_conflicts


def _walk_rel_paths(repo: Path) -> list[str]:
    out: list[str] = []
    skip = {".git", "node_modules", ".tonic"}

    def walk(base: Path, rel: str) -> None:
        try:
            for e in base.iterdir():
                if e.name in skip:
                    continue
                r = f"{rel}/{e.name}".strip("/") if rel else e.name
                r = r.replace("\\", "/")
                if e.is_dir():
                    walk(e, r)
                elif e.is_file():
                    out.append(r)
        except OSError:
            return

    walk(repo, "")
    return sorted(out)


def scan_repo_conflict_markers(repo: Path) -> dict[str, Any]:
    regions: list[dict[str, Any]] = []
    rid = 0
    for rel in _walk_rel_paths(repo):
        parts = [x for x in rel.replace("\\", "/").split("/") if x]
        abs_p = repo.joinpath(*parts) if parts else repo
        try:
            text = abs_p.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue
        blocks = parse_git_conflicts(text)
        for b in blocks:
            rid += 1
            start_1 = b.start_line + 1
            end_1 = b.end_line + 1
            mid = (start_1 + end_1) // 2
            ours = b.segments[0].label if b.segments else ""
            theirs = b.segments[1].label if len(b.segments) > 1 else ""
            row: dict[str, Any] = {
                "path": rel.replace("\\", "/"),
                "region_id": f"r{rid}",
                "start_line": start_1,
                "mid_line": mid,
                "end_line": end_1,
            }
            if ours:
                row["ours_label"] = ours
            if theirs:
                row["theirs_label"] = theirs
            regions.append(row)
    regions.sort(key=lambda x: (x["path"], x["start_line"]))
    return {
        "schema": "tonic-conflict-context",
        "version": "1",
        "scan_scope": "workspace",
        "conflict_regions": regions,
    }


def write_conflict_context(path_out: str, art: dict[str, Any]) -> None:
    p = Path(path_out).resolve()
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(art, indent=2) + "\n", encoding="utf-8")

File: packages/tonic-core/src/hydration/conflictScan.ts
import * as fs from "node:fs";
import * as path from "node:path";

import { parseGitConflictsWithDiagnostics } from "../gitConflictParser";

export type ConflictRegionV1 = {
  path: string;
  region_id?: string;
  start_line: number;
  mid_line: number;
  end_line: number;
  ours_label?: string;
  theirs_label?: string;
  hunk_id?: string;
};

export type ConflictContextArtifactV1 = {
  schema: "tonic-conflict-context";
  version: "1";
  scan_scope: string;
  conflict_regions: ConflictRegionV1[];
};

function walkFiles(repoRoot: string): string[] {
  const out: string[] = [];
  function walk(dir: string, base: string): void {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const e of entries) {
      if (e.name === ".git" || e.name === "node_modules" || e.name === ".tonic") {
        continue;
      }
      const rel = path.join(base, e.name).replace(/\\/g, "/");
      const full = path.join(dir, e.name);
      if (e.isDirectory()) {
        walk(full, rel);
      } else if (e.isFile()) {
        out.push(rel);
      }
    }
  }
  walk(repoRoot, "");
  return out.sort();
}

export function scanRepoConflictMarkers(repoRoot: string): ConflictContextArtifactV1 {
  const relPaths = walkFiles(repoRoot);
  const regions: ConflictRegionV1[] = [];
  let rid = 0;
  for (const rel of relPaths) {
    const abs = path.join(repoRoot, rel);
    let text: string;
    try {
      text = fs.readFileSync(abs, "utf8");
    } catch {
      continue;
    }
    const { blocks } = parseGitConflictsWithDiagnostics(text);
    for (const b of blocks) {
      rid += 1;
      const startLine = b.startLine + 1;
      const endLine = b.endLine + 1;
      const midLine = Math.floor((startLine + endLine) / 2);
      const ours = b.segments[0]?.label ?? "";
      const theirs = b.segments[1]?.label ?? "";
      regions.push({
        path: rel.replace(/\\/g, "/"),
        region_id: `r${rid}`,
        start_line: startLine,
        mid_line: midLine,
        end_line: endLine,
        ours_label: ours || undefined,
        theirs_label: theirs || undefined,
      });
    }
  }
  regions.sort((a, b) => (a.path !== b.path ? a.path.localeCompare(b.path) : a.start_line - b.start_line));
  return {
    schema: "tonic-conflict-context",
    version: "1",
    scan_scope: "workspace",
    conflict_regions: regions,
  };
}

export function writeConflictContext(pathOut: string, art: ConflictContextArtifactV1): void {
  fs.mkdirSync(path.dirname(path.resolve(pathOut)), { recursive: true });
  fs.writeFileSync(pathOut, JSON.stringify(art, null, 2) + "\n", "utf8");
}
